docs/tutorials/sparse/csr.md

```diff
@@ -0,0 +1,258 @@
+# CSRNDArray - NDArray in Compressed Sparse Row Storage Format
+
+Many real world datasets deal with high dimensional sparse feature vectors. For instance,
+in a recommendation system, the number of categories and users is in the order of millions,
+while most users only made a few purchases, leading to feature vectors with high sparsity
+(i.e. most of the elements are zeros).
+
+Storing and manipulating such large sparse matrices in the default dense structure results
+in wasted memory and processing on the zeros.
+To take advantage of the sparse structure of the matrix, the ``CSRNDArray`` in MXNet
+stores the matrix in [compressed sparse row(CSR)](https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_.28CSR.2C_CRS_or_Yale_format.29) format
+and uses specialized algorithms in operators.
+The format is designed for 2D matrices with a large number of columns,
+and each row is sparse(i.e. with only a few nonzeros).
+For matrices of high sparsity (e.g. ~1% non-zeros), the advantage of ``CSRNDArray`` over
+the existing ``NDArray`` is that
+
+- memory consumption is reduced significantly
+- certain operations (e.g. matrix-vector multiplication) are much faster
+
+Meanwhile, ``CSRNDArray`` inherits competitive features from ``NDArray`` such as
+lazy evaluation and automatic parallelization, which are not available in the
+scientific computing python package [SciPy](https://www.scipy.org/).
+
+Apart from often queried attributes such as **ndarray.shape**, **ndarray.dtype** and **ndarray.context**,
+you’ll also want to query **ndarray.stype**: the storage type of the NDArray. For a usual dense NDArray,
+the value of stype is **"default"**. For an CSRNDArray, the value of stype is **"csr"**.
+
+## Prerequisites
+
+To complete this tutorial, we need:
+
+- MXNet. See the instructions for your operating system in [Setup and Installation](http://mxnet.io/get_started/install.html)
+- [Jupyter](http://jupyter.org/)
+    ```
+    pip install jupyter
+    ```
+- Scipy - A section of this tutorial uses Scipy package in python. If you don't have Scipy,
+the example in that section will be ignored.
+- GPUs - A section of this tutorial uses GPUs. If you don't have GPUs on your
+machine, simply set the variable gpu_device (set in the GPUs section of this
+tutorial) to mx.cpu().
+
+## Compressed Sparse Row Format
+
+A CSRNDArray represents a 2D matrix as three separate 1D arrays: **data**,
+**indptr** and **indices**, where the column indices for
+row ``i`` are stored in ``indices[indptr[i]:indptr[i+1]]`` in ascending order,
+and their corresponding values are stored in ``data[indptr[i]:indptr[i+1]]``.
+
+For example, the CSR representation for matrix
+```
+[[7, 0, 8, 0]
+ [0, 0, 0, 0]
+ [0, 9, 0, 0]]
+```
+is:
+```
+[7, 8, 9]          # data
+[0, 2, 1]          # indices
+[0, 2, 2, 3]       # indptr
+```
+
```

Comment:
Suggestion: I think the suggested text below may help newbies understand the various numbers better. Try it with a newbie if you like (correct the text spacing appropriately).

[7, 8, 9]          # data: flattened representation of the dense matrix in row-major format after removing all zeros.
[0, 2, 1]          # indices: column indices pointing to the non-zero elements in the dense matrix.
[0, 2, 2, 3]     # indptr: index pointers into data[] array that signify start of a row in the dense matrix.
                      # i.e. Row 0 starts at index pointer 0, pointing to element 7, in data[].
                      # i.e. Row 1 starts at index pointer 2, pointing to element 9, in data[] since Row 1 is all-zeroes.
                      # i.e. Row 2 starts at index pointer 2, pointing to element 9, in data[].
                      # i.e. the last element in indptr is always one past the size of data[], signify end of data[].